core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java

```diff
@@ -194,184 +197,278 @@ public ClusterManagerStatus getStatus() {
+      }
+      throw CompletableFutures.asCompletionException(e);
+   }
+
+   private ClusterCacheStatus prepareJoin(String cacheName, Address joiner, CacheJoinInfo joinInfo,
+                                          int joinerViewId) {
+      updateLock.lock();
+      try {
          if (!clusterManagerStatus.isRunning()) {
             log.debugf("Ignoring join request from %s for cache %s, the local cache manager is shutting down",
-                  joiner, cacheName);
-            return null;
+                       joiner, cacheName);
+            throw new IllegalLifecycleStateException();
          }
          if (joinerViewId < viewId) {
-            log.debugf("Ignoring join request from %s for cache %s, joiner's view id is too old: %d", joiner,
-                  cacheName, joinerViewId);
+            log.debugf("Ignoring join request from %s for cache %s, joiner's view id is too old: %d",
+                       joiner, cacheName, joinerViewId);
             return null;
          }
 
-         cacheStatus = initCacheStatusIfAbsent(cacheName, joinInfo.getCacheMode());
+         return initCacheStatusIfAbsent(cacheName, joinInfo.getCacheMode());
+      } finally {
+         updateLock.unlock();
+      }
+   }
+
+   private boolean canHandleJoin(int joinerViewId) {
+      updateLock.lock();
+      try {
+         return joinerViewId <= viewId &&
+                clusterManagerStatus != ClusterManagerStatus.RECOVERING_CLUSTER &&
+                clusterManagerStatus != ClusterManagerStatus.INITIALIZING;
       } finally {
-         clusterManagerLock.unlock();
+         updateLock.unlock();
       }
-      return cacheStatus.doJoin(joiner, joinInfo);
    }
 
    @Override
-   public void handleLeave(String cacheName, Address leaver, int viewId) throws Exception {
+   public CompletionStage<Void> handleLeave(String cacheName, Address leaver, int viewId) throws Exception {
       if (!clusterManagerStatus.isRunning()) {
          log.debugf("Ignoring leave request from %s for cache %s, the local cache manager is shutting down",
-               leaver, cacheName);
-         return;
+                    leaver, cacheName);
+         return CompletableFutures.completedNull();
       }
 
       ClusterCacheStatus cacheStatus = cacheStatusMap.get(cacheName);
       if (cacheStatus == null) {
          // This can happen if we've just become coordinator
-         log.tracef("Ignoring leave request from %s for cache %s because it doesn't have a cache status entry", leaver, cacheName);
-         return;
-      }
-      if (cacheStatus.doLeave(leaver)) {
-         cacheStatusMap.remove(cacheName);
+         log.tracef("Ignoring leave request from %s for cache %s because it doesn't have a cache status entry",
+                    leaver, cacheName);
+         return CompletableFutures.completedNull();
       }
+      return cacheStatus.doLeave(leaver);
+   }
+
+   synchronized void removeCacheStatus(String cacheName) {
```

Comment:
this isn't invoked anywhere. 
should it be invoked after the handleLeave?

Reply:
It's invoked by `ClusterCacheStatus.doLeave()`